jina/hubble/hubio.py

```diff
@@ -476,88 +478,147 @@ def push(self) -> None:
                     form_data['secret'] = self.args.secret or secret
 
                 st.update(f'Connecting to Jina Hub ...')
-                if form_data.get('id'):
-                    hubble_url = urljoin(hubble.utils.get_base_url(), 'executor.update')
-                else:
-                    hubble_url = urljoin(hubble.utils.get_base_url(), 'executor.create')
-
-                # upload the archived executor to Jina Hub
-                st.update(f'Uploading...')
-                resp = upload_file(
-                    hubble_url,
-                    'filename',
-                    content,
-                    dict_data=form_data,
-                    headers=req_header,
-                    stream=True,
-                    method='post',
-                )
 
-                image = None
-                warning = None
-                session_id = req_header.get('jinameta-session-id')
-                for stream_line in resp.iter_lines():
-                    stream_msg = json.loads(stream_line)
-
-                    t = stream_msg.get('type')
-                    subject = stream_msg.get('subject')
-                    payload = stream_msg.get('payload', '')
-                    if t == 'error':
-                        msg = stream_msg.get('message')
-                        hubble_err = payload
-                        overridden_msg = ''
-                        detail_msg = ''
-                        if isinstance(hubble_err, dict):
-                            (overridden_msg, detail_msg) = get_hubble_error_message(
-                                hubble_err
+                logged_in = True if is_logged_in() else False;
+                # if user logged in use async upload api
+                if logged_in: 
+
+                    if task_id and not form_data.get('id'):
+                        image = self._query_with_progress(task_id)
+
+                        if image: 
+                            new_uuid8, new_secret = self._prettyprint_result(
+                                console, image, warning=False
                             )
-                            if not msg:
-                                msg = detail_msg
 
-                        if overridden_msg and overridden_msg != detail_msg:
-                            self.logger.warning(overridden_msg)
+                            if new_uuid8 != uuid8 or new_secret != secret:
+                                dump_secret(work_path, new_uuid8, new_secret or '', task_id)
+                        return
+
+                    
+                    if form_data.get('id'):
+                        hubble_url = urljoin(hubble.utils.get_base_url(), 'executor.updateAsync') 
+                    else:
+                        hubble_url = urljoin(hubble.utils.get_base_url(), 'executor.createAsync')
+
+                    # upload the archived executor to Jina Hub
+                    st.update(f'Async Uploading...')
```

Comment:
how is this async?

Reply:
mm, we use async API, the user does not have to wait for the build process to create or update the executor, we will put it in the task queue for execution